### PR TITLE
Draft: Retrievers

### PR DESCRIPTION
**Pull request by tomlynchRNA**
#151

### CODE REVIEW

File: agent-backend/src/build/build_crew.py [hunk 1]
```diff
@@ -191,8 +108,12 @@ def build_tools_and_their_datasources(self):
             for model_key, model_object in tool_models_objects.items():
                 model_data = match_key(self.models_models, model_key)
                 tool_models_models.append((model_object, model_data))
+
+            # Use same llm as Crew's for retriever
+            retriever_llm = match_key(self.crew_models, keyset(self.crew_model.id))
```
Comment (charl3sj) [inline agent-backend/src/build/build_crew.py:113]: The self-query retriever needs an LLM... for now I decided to pass the same LLM linked to the Crew. But we have to either take it as a user input or use one of the already configured models linked to Agent or Crew (as it is now). Let me know your thoughts!

Comment (tomlynchRNA) [reply agent-backend/src/build/build_crew.py:113]: For now yes, the crew or "manager" llm is fine :+1:

File: agent-backend/src/tools/retrievers/self_query.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+class SelfQueryRetriever(BaseToolRetriever):
+    def __init__(self, tool: Tool, llm: BaseLanguageModel, vector_store: VectorStore):
+        metadata_field_info = list(
+            map(lambda x: AttributeInfo(**x.model_dump()), tool.retriever_config.metadata_field_info))
+
+        # TODO: Change the `document_contents` param value from `tool.description` to something else
+        self.retriever = LC_SelfQueryRetriever.from_llm(llm=llm,
+                                                        vectorstore=vector_store,
+                                                        document_contents=tool.description,
```
Comment (charl3sj) [inline agent-backend/src/tools/retrievers/self_query.py:18]: This parameter is meant to describe the contents of the document being queried. Should we introduce a new sub-field for this or refer to the datasource's description?

Comment (tomlynchRNA) [reply agent-backend/src/tools/retrievers/self_query.py:18]: tool description is correct

**Pull request merged**